### PR TITLE
feat: Stoppable

### PR DESCRIPTION
**Pull request by Eknir**
This functionality makes the StorageManager stoppable by the owner of the contract (by default the person who deploys the contract). Certain functions, chosen such that existing agreements can continue but no new agreements/changes to existing agreements can be made, cannot be executed anymore after the contract is stopped. This adds the possibility of upgrading.

The contract Stoppable is largely based on the openzeppelin [Pausable](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/Pausable.sol) standard, removing the functionality to re-start the contract.

### CODE REVIEW

**Comment by julianmrodri** [comment]
@Eknir why not use `Pausable` directly? Also why not allow to restart/resume the contract? (i think this is part of the bigger discussion if we are going for Upgradability or not) But if this is going to be Upgradable we definitely need the `unpause` (restart).

File: package.json [hunk 1]
```diff
@@ -36,8 +36,11 @@
     "release": "tasegir release --no-docs --no-compile --no-test --no-build",
     "prepublishOnly": "truffle compile && typechain --target=web3-v1 build/contracts/StorageManager.json && typechain --target=truffle-v4 --outDir ./types/truffle-v4-contracts build/contracts/StorageManager.json && typechain --target=truffle-v5 --outDir ./types/truffle-v5-contracts build/contracts/StorageManager.json && typechain --target=ethers-v4 build/contracts/StorageManager.json"
   },
-  "dependencies": {},
+  "dependencies": {
+    "@openzeppelin/test-environment": "^0.1.4"
```
Comment (AuHau) [inline package.json:40]: This should be `devDependency`, but I am not sure why you even need this dependency?

**Comment by vojtechsimetka** [comment]
> @Eknir why not use `Pausable` directly? Also why not allow to restart/resume the contract? (i think this is part of the bigger discussion if we are going for Upgradability or not) But if this is going to be Upgradable we definitely need the `unpause` (restart).

@julianmrodri This was my decision after a brief discussion with @Eknir . We decided the simplest solution is to do a one way operation as any upgradability here means redeploying the contracts anyways. Unpausing could cause all sorts of issues and could even be triggered on accident for some past version of a contract.

**Comment by julianmrodri** [comment]
@vojtechsimetka Why do you say upgradability would require to redelpoy the contracts? Im missing that. If thats the approach taken then with much more reason Staking should be outside of this contract so it persists independent of the actual service. (related to the other PR)

File: package.json (hunk 1 above)
Comment (Eknir) [reply package.json]: Put it to `devDependency`. I'm using it in `stoppable.test.js`

**Comment by Eknir** [comment]
Hi guys, thanks for the feedback.

About upgradability: @julianmrodri , there are many patters to make contracts "upgradable", most of them are questionable whether you actually want them or not. The easiest pattern to make contracts upgradable is to actually stop the functionality of one contract, at the moment that you deployed a new contract. This way, you can force anybody who implemented a front-end to the contract to update the reference to the "old" smart-contract and work from that moment onwards with the new smart-contract.

Depending on your solution, this could actually work very well to upgrade smart-contracts, it is easy to understand and easy to implement. If you want another way of upgrading smart-contracts I would recommend that you exactly spell out what when and why you need to upgrade your contracts. That way, we can find the best pattern possible (but I expect that this pattern would work anyways).

About modifying `Pausable` to be `Stoppable`: a contract must be stopped to be up gradable with the pattern I just described. If we make the contract `Pausable`, users might still expect that at some point, the smart-contract gets updated. We explicitly don't want that, as we want to stop the old contract. Of course, using Pausable would work too, but if you do this, you need to have good practices to also destroy any reference to the private key with which the original contract was deployed. I think it is more hassle.

**Comment by vojtechsimetka** [comment]
Thanks @Eknir for sharing that reasoning, I would not write it better 😁 

@julianmrodri maybe you can share what is the upgrade mechanism for the RNS marketplace contracts?

**Comment by Eknir** [comment]
Closing, as we agreed we will go with another mechanism to upgrade contracts

**Pull request closed without merging**